runtime/pkg/rilltime/rilltime.go

```diff
@@ -271,277 +338,303 @@ func (e *Expression) Eval(evalOpts EvalOptions) (time.Time, time.Time, timeutil.
+	} else if i.Iso != nil {
+		return i.Iso.eval(tz)
+	}
+	return start, start, timeutil.TimeGrainUnspecified
+}
 
-	i := len(l.anchorParts) - 1
-	for i >= 0 {
-		start, cur, end = l.anchorParts[i].eval(evalOpts, start, cur, end, tz)
-		i--
-	}
+func (o *AnchoredDurationInterval) eval(evalOpts EvalOptions, tm time.Time, tz *time.Location) (time.Time, time.Time, timeutil.TimeGrain) {
+	start, _ := o.PointInTime.eval(evalOpts, tm, tz)
+	end := start
 
 	tg := timeutil.TimeGrainUnspecified
-	if len(l.anchorParts) > 0 {
-		tg = l.anchorParts[0].grain()
+	if o.Starting {
+		end, tg = o.Duration.offset(start, 1)
+	} else if o.Ending {
+		start, tg = o.Duration.offset(end, -1)
 	}
 
 	return start, end, tg
 }
 
-func (l *LinkPart) parse() (linkPartAnchor, error) {
-	var lpa linkPartAnchor
-	if l.Ordinal != nil {
-		lpa = l.Ordinal
-	} else if l.Anchor != nil {
-		lpa = l.Anchor
-	} else if l.AbsoluteTime != nil {
-		lpa = l.AbsoluteTime
-	} else if l.LabeledAnchor != nil {
-		lpa = l.LabeledAnchor
+func (o *OrdinalInterval) eval(evalOpts EvalOptions, start time.Time, tz *time.Location) (time.Time, time.Time, timeutil.TimeGrain) {
+	end := start
+	if o.End != nil {
+		start, end, _ = o.End.eval(evalOpts, start, tz)
+	} else if len(o.Ordinal.Durations) > 0 {
+		lastPart := o.Ordinal.Durations[len(o.Ordinal.Durations)-1]
+		if lastPart.Ordinal != nil {
+			tg := higherOrderMap[grainMap[lastPart.Ordinal.Grain]]
+			start = truncateWithCorrection(start, tg, tz, evalOpts.FirstDay, evalOpts.FirstMonth)
+			end = timeutil.CeilTime(end, tg, tz, evalOpts.FirstDay, evalOpts.FirstMonth)
+		}
 	}
 
-	if lpa == nil {
-		return nil, fmt.Errorf("invalid link part: atleast one of ordinal, anchor, absolute time or labeled anchor must be specified")
+	start, end, tg := o.Ordinal.eval(evalOpts, start, end, tz)
+
+	return start, end, tg
+}
+
+func (o *OrdinalIntervalEnd) eval(evalOpts EvalOptions, start time.Time, tz *time.Location) (time.Time, time.Time, timeutil.TimeGrain) {
+	if o.Grains != nil {
+		return o.Grains.eval(evalOpts, start, tz)
+	} else if o.Interval != nil {
+		return o.Interval.eval(evalOpts, start, tz)
+	} else if o.SingleGrain != nil {
+		tg := grainMap[*o.SingleGrain]
+		end := timeutil.CeilTime(start, tg, tz, evalOpts.FirstDay, evalOpts.FirstMonth)
+		start = truncateWithCorrection(start, tg, tz, evalOpts.FirstDay, evalOpts.FirstMonth)
+		return start, end, tg
 	}
+	return start, start, timeutil.TimeGrainUnspecified
+}
 
-	err := lpa.parse()
-	if err != nil {
-		return nil, err
+func (o *StartEndInterval) eval(evalOpts EvalOptions, tm time.Time, tz *time.Location) (time.Time, time.Time, timeutil.TimeGrain) {
+	start, startTg := o.Start.eval(evalOpts, tm, tz)
+	end, endTg := o.End.eval(evalOpts, tm, tz)
+	// Correction for labeled ends points. We need to add +1ms to make sure the final point is included as our end time is exclusive.
+	if o.End.Labeled != nil {
+		end = timeutil.OffsetTime(end, timeutil.TimeGrainMillisecond, 1)
```

Comment:
This really isn't safe – we don't know what the smallest time granularity is. It can vary between nanosecond and second granularity depending on the timestamp type used.